// react-multistep/dist/index.js
'use strict';

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = require('react');
var React__default = _interopDefault(React);

function styleInject(css, ref) {
  if ( ref === void 0 ) ref = {};
  var insertAt = ref.insertAt;

  if (!css || typeof document === 'undefined') { return; }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css = "ol {\n  margin: 0;\n  padding-bottom: 2.2rem;\n  list-style-type: none;\n}\nli {\n  display: inline-block;\n  text-align: center;\n  line-height: 4.5rem;\n  padding: 0 0.7rem;\n  cursor: pointer;\n}\nli:after {\n  content: \"\\A0\\A0\";\n}\nli:before {\n  position: relative;\n  bottom: -3.7rem;\n  float: left;\n  left: 50%;\n}\nli span {\n  padding: 0 1.5rem;\n}\n@media (max-width: 650px) {\n  li span {\n    display: none;\n  }\n}\n\n.index_progtrckr-todo__2rCwX {\n  color: silver;\n  border-bottom: 4px solid silver;\n}\n.index_progtrckr-todo__2rCwX:before {\n  content: \"\\39F\";\n  color: silver;\n  background-color: white;\n  width: 1.2em;\n  line-height: 1.4em;\n}\n.index_progtrckr-todo__2rCwX:hover:before {\n  color: #0FA0CE;\n}\n\n\n.index_progtrckr-doing__3GALV {\n  color: black;\n  border-bottom: 4px solid #33C3F0;\n}\n.index_progtrckr-doing__3GALV:before {\n  content: \"\\2022\";\n  color: white;\n  background-color: #33C3F0;\n  width: 1.2em;\n  line-height: 1.2em;\n  border-radius: 1.2em;\n}\n.index_progtrckr-doing__3GALV:hover:before {\n  color: #0FA0CE;\n}\n\n.index_progtrckr-done__c0N9Q {\n  color: black;\n  border-bottom: 4px solid #33C3F0;\n}\n.index_progtrckr-done__c0N9Q:before {\n  content: \"\\2713\";\n  color: white;\n  background-color: #33C3F0;\n  width: 1.2em;\n  line-height: 1.2em;\n  border-radius: 1.2em;\n}\n.index_progtrckr-done__c0N9Q:hover:before {\n  color: #0FA0CE;\n}\n";
var css$1 = { "progtrckr-todo": "index_progtrckr-todo__2rCwX", "progtrckr-doing": "index_progtrckr-doing__3GALV", "progtrckr-done": "index_progtrckr-done__c0N9Q" };
styleInject(css);

var slicedToArray = function () {
  function sliceIterator(arr, i) {
    var _arr = [];
    var _n = true;
    var _d = false;
    var _e = undefined;

    try {
      for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) {
        _arr.push(_s.value);

        if (i && _arr.length === i) break;
      }
    } catch (err) {
      _d = true;
      _e = err;
    } finally {
      try {
        if (!_n && _i["return"]) _i["return"]();
      } finally {
        if (_d) throw _e;
      }
    }

    return _arr;
  }

  return function (arr, i) {
    if (Array.isArray(arr)) {
      return arr;
    } else if (Symbol.iterator in Object(arr)) {
      return sliceIterator(arr, i);
    } else {
      throw new TypeError("Invalid attempt to destructure non-iterable instance");
    }
  };
}();

var getNavStyles = function getNavStyles(indx, length) {
  var styles = [];
  for (var i = 0; i < length; i++) {
    if (i < indx) {
      styles.push('done');
    } else if (i === indx) {
      styles.push('doing');
    } else {
      styles.push('todo');
    }
  }
  return styles;
};

var getButtonsState = function getButtonsState(indx, length) {
  if (indx > 0 && indx < length - 1) {
    return {
      showPreviousBtn: true,
      showNextBtn: true
    };
  } else if (indx === 0) {
    return {
      showPreviousBtn: false,
      showNextBtn: true
    };
  } else {
    return {
      showPreviousBtn: true,
      showNextBtn: false
    };
  }
};

function MultiStep(props) {
  var showNavigation = true;
  if (props.showNavigation && props.showNavigation) showNavigation = props.showNavigation;

  var _useState = React.useState(getNavStyles(0, props.steps.length)),
      _useState2 = slicedToArray(_useState, 2),
      stylesState = _useState2[0],
      setStyles = _useState2[1];

  var _useState3 = React.useState(0),
      _useState4 = slicedToArray(_useState3, 2),
      compState = _useState4[0],
      setComp = _useState4[1];

  var _useState5 = React.useState(getButtonsState(0, props.steps.length)),
      _useState6 = slicedToArray(_useState5, 2),
      buttonsState = _useState6[0],
      setButtons = _useState6[1];

  function setStepState(indx) {
    setStyles(getNavStyles(indx, props.steps.length));
    setComp(indx < props.steps.length ? indx : compState);
    setButtons(getButtonsState(indx, props.steps.length));
  }

  var next = function next() {
    return setStepState(compState + 1);
  };

  var previous = function previous() {
    return setStepState(compState > 0 ? compState - 1 : compState);
  };

  var handleKeyDown = function handleKeyDown(evt) {
    return evt.which === 13 ? next(props.steps.length) : {};
  };

  var handleOnClick = function handleOnClick(evt) {
    if (evt.currentTarget.value === props.steps.length - 1 && compState === props.steps.length - 1) {
      setStepState(props.steps.length);
    } else {
      setStepState(evt.currentTarget.value);
    }
  };

  var renderSteps = function renderSteps() {
    return props.steps.map(function (s, i) {
      return React__default.createElement(
        'li',
        {
          className: css$1['progtrckr-' + stylesState[i]],
          onClick: handleOnClick,
          key: i,
          value: i
        },
        React__default.createElement(
          'span',
          null,
          props.steps[i].name
        )
      );
    });
  };

  var renderNav = function renderNav(show) {
    return show && React__default.createElement(
      'div',
      null,
      React__default.createElement(
        'button',
        {
          style: buttonsState.showPreviousBtn ? {} : { display: 'none' },
          onClick: previous
        },
        'Prev'
      ),
      React__default.createElement(
        'button',
        {
          style: buttonsState.showNextBtn ? {} : { display: 'none' },
          onClick: next
        },
        'Next'
      )
    );
  };

  return React__default.createElement(
    'div',
    { onKeyDown: handleKeyDown },
    React__default.createElement(
      'ol',
      null,
      renderSteps()
    ),
    React__default.createElement(
      'div',
      null,
      props.steps[compState].component
    ),
    React__default.createElement(
      'div',
      null,
      renderNav(showNavigation)
    )
  );
}

module.exports = MultiStep;
